### ISSUE TITLE
using ignition kargs features loses transient firstboot kernel arguments

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#882]
**Describe the bug**

If I use the ignition kernel argument [new feature](https://docs.fedoraproject.org/en-US/fedora-coreos/kernel-args/#_modifying_kernel_arguments_via_ignition) I lose any transient kernel arguments between the firstboot (when kargs are applied) and the second boot (when the rest of Ignition and full boot into real root occurs).

In this case I'm defining transient kernel arguments as:
- kernel arguments that were applied on first boot (i.e. PXE, typed by hand, fed into a hypervisor, etc...) but are not otherwise a part of the set of kernel arguments delivered by the OS and don't make sense or don't need to be permanently applied.

There are a host of small corner cases that makes this not ideal. Some of them off the top of my head:

- first boot networking kargs are lost
- `ignition.config.url` gets lost

**Reproduction steps**
1. Create Ignition config with kernel argument modifications (see https://docs.fedoraproject.org/en-US/fedora-coreos/kernel-args/#_modifying_kernel_arguments_via_ignition) and an SSH key.
2. Host Ignition config on web server somewhere
3. Boot machine and type `ignition.config.url=https://xyz.com`
4. Observe the final machine has the following on the console:

```
Ignition: no config provided by user
No ssh authorized keys provided by Ignition or Afterburn
```

**Expected behavior**
Transient arguments get carried through. 

**Actual behavior**
Loss of transient arguments

**System details**
 - QEMU with `ignition.config.url` and no fwcfg
 - `34.20210529.3.0`

**Ignition config**

Something like this butane config should do it:

```
variant: fcos
version: 1.4.0-experimental
kernel_arguments:
  should_exist:
    - mitigations=off
  should_not_exist:
    - mitigations=auto,nosmt
passwd:
  users:
    - name: core
      ssh_authorized_keys:
        - $sshpubkey
```

**Comment by bgilbert** [comment]
Good catch.  Hand-typed kargs are the obvious case here.  One solution is for the kargs helper script to notice any unexpected kargs in `/proc/cmdline` and copy them back into the `ignition.firstboot` file.  That doesn't require any Ignition changes, but we'd need to document upstream that the distro kargs script needs to preserve first-boot kargs if the distro cares about doing that.

PXE is a very different case.  With live PXE, the PXE server completely controls the kargs on every boot, and there's not even a `/boot` filesystem for us to update.  So I'd assume the kargs helper will just fail outright if the `kernelArguments` section is used in that case.

For "fed into a hypervisor", do you mean e.g. direct kernel booting with qemu?  We've usually treated that case as equivalent to live PXE, since it bypasses the bootloader configs that would normally drive OS updates.

**Comment by dustymabe** [comment]
> Good catch. Hand-typed kargs are the obvious case here. One solution is for the kargs helper script to notice any unexpected kargs in `/proc/cmdline` and copy them back into the `ignition.firstboot` file. That doesn't require any Ignition changes, but we'd need to document upstream that the distro kargs script needs to preserve first-boot kargs if the distro cares about doing that.

Exactly what I was thinking. Basicallly detect what args exist that weren't provided by the distro and propagate those into that next boot.

> 
> PXE is a very different case. With live PXE, the PXE server completely controls the kargs on every boot, and there's not even a `/boot` filesystem for us to update. So I'd assume the kargs helper will just fail outright if the `kernelArguments` section is used in that case.

Yep. In the LIVE PXE case there's nothing for us to do. The kernel arguments are defined on every boot (i.e. the source for the kernel arguments don't come from the disk) by the PXE config. There's not even a /boot partition IIUC so we couldn't even if we wanted to.

The case here that is affected is "PXE install". So running the Live PXE environment to run `coreos-installer`, which auto propagates networking kargs forward into the ignition boot. That boot will now apply the Ignition kargs and reboot, losing the forwarded networking kargs.

> 
> For "fed into a hypervisor", do you mean e.g. direct kernel booting with qemu? We've usually treated that case as equivalent to live PXE, since it bypasses the bootloader configs that would normally drive OS updates.

Right, it is like Live PXE unless you use `virt-install --install` which will use the disk provided kernel/initrd/args after the initial boot. https://dustymabe.com/2020/01/30/virt-install-boot-from-specific-kernel/initrd-just-for-install/

**Comment by bgilbert** [comment]
> So running the Live PXE environment to run coreos-installer, which auto propagates networking kargs forward into the ignition boot. That boot will now apply the Ignition kargs and reboot, losing the forwarded networking kargs.

Hmm, I wasn't able to reproduce this behavior.  Do you have a reproducer?

> Right, it is like Live PXE unless you use virt-install --install which will use the disk provided kernel/initrd/args after the initial boot.

Ah, okay, so this should reduce to the PXE install case.

**Comment by dustymabe** [comment]
> > So running the Live PXE environment to run coreos-installer, which auto propagates networking kargs forward into the ignition boot. That boot will now apply the Ignition kargs and reboot, losing the forwarded networking kargs.
> 
> Hmm, I wasn't able to reproduce this behavior. Do you have a reproducer?

Ahh yes. That's because they're backed by a file https://github.com/coreos/coreos-installer/blob/c4fabc7c801491019263523df3c9078cca99bb6e/src/install.rs#L270-L289. Sorry for the noise, I assumed this case was affected. 

> 
> > Right, it is like Live PXE unless you use virt-install --install which will use the disk provided kernel/initrd/args after the initial boot.
> 
> Ah, okay, so this should reduce to the PXE install case.

Correct. Though, hmm. Since the `coreos-installer` properly preserves the arguments in `ignition.firstboot` file then the "PXE install"  case is also not affected.

I guess the *only* thing we should be concerned about is the "hand typed" case, but most of the time people are doing that they would be doing it for `coreos-installer`, right?

Where I hit this was running the [manual network testing script](https://github.com/coreos/fedora-coreos-config/blob/fffae203a834fc325cbd8e6ddd8a585e2d9df21d/tests/manual/coreos-network-testing.sh) which extracts the kernel/initrd from the qcow2 so that it can then start the machine with whatever kernel arguments it wants to (for the first boot only). This is convenient for testing initrd networking scenarios, but it does simulate "hand typed kernel arguments". When I added in kernel arguments in Ignition (to test something else) I hit this case where the networking kargs were lost on that second boot with ignition kargs applied, hence filing this issue.

**Comment by bgilbert** [comment]
> I guess the _only_ thing we should be concerned about is the "hand typed" case, but most of the time people are doing that they would be doing it for `coreos-installer`, right?

Probably, yeah.  To my mind, the biggest gap is with users deploying a one-off cloud image (without using coreos-installer) and hand-typing `ignition.config.url`.  I don't know if that happens very much, though, and it looks like we don't document that flow.

**Comment by jlebon** [comment]
I'm not confident we should automatically handle hand-typed kargs. If you've typed kargs in, then you know that it's not going to last the next boot. For each automatic reboot we have or could introduce in the future, we would have to make a judgement call about whether the user still wants the hand-typed kargs applied or not. In the case of Ignition kargs, it's more (but not entirely) obvious, but I'd rather just stay out of it entirely. (Is the real root the threshold? What if it's some kernel arg that they really meant to only apply on the first boot?)

Hmm, I guess if they typed things in, then they're at the console and they could just retype it in on reboot if that's what they want.

**Comment by dustymabe** [comment]
> I'm not confident we should automatically handle hand-typed kargs. If you've typed kargs in, then you know that it's not going to last the next boot.

Yes, but we've shifted now the definition of "next boot".

> For each automatic reboot we have or could introduce in the future, we would have to make a judgement call about whether the user still wants the hand-typed kargs applied or not. In the case of Ignition kargs, it's more (but not entirely) obvious, but I'd rather just stay out of it entirely. (Is the real root the threshold? What if it's some kernel arg that they really meant to only apply on the first boot?)

Yeah I think the real root would be the threshold. Basically if we reboot before the real root then you can almost consider like we never actually booted and it's still "first boot". In that case then yes, the user wants the arg on the "next boot" that they meant to only apply on "first boot". Otherwise they would have included it in the Ignition config. One contrived example I can think of is a user using `enforcing=0` on first boot and then using a systemd unit (runs in the real root) written by ignition to fix up labels on some files. 

The way I look at it is trying to get parity with what we had before. I guess we could flip the conversation and ask: 

- *what problems do we think can arise from capturing and adding those transient args to the next boot?*

Are there big problems? Are the users less or more surprised than the current behavior?

> 
> Hmm, I guess if they typed things in, then they're at the console and they could just retype it in on reboot if that's what they want.

Yeah, though I think that's a bad UX. Imagine having to catch the grub prompt twice!

**Comment by bgilbert** [comment]
Amusingly, if the user is hand-typing `ignition.config.url=https://example.com/config.ign`, then one workaround is:

```yaml
variant: fcos
version: 1.4.0
kernel_arguments:
  should_exist:
    - ignition.config.url=https://example.com/config.ign
```

Overall, I agree that this behavior is surprising, but I'm not sure there's a strong case for prioritizing a fix.  I'd be inclined to leave this bug open for a while and see what use cases it accumulates.
